content/posts/deploying-application-on-k8s-with-helm/index.md

```diff
@@ -0,0 +1,431 @@
+
+I will introduce you to the structure of the helm chart. To create a chart, we can use my example command:
+```bash
+
+  helm create mynginx
+
+```
+That command will create an mynginx folder. Let take a look in mynginx folder, we have some folders and files:
+- charts: this folder will contain another subchart which our chart will depend on. I will talk more about subchart it in the next section
+- templates: this folder contains yaml files about the k8s resource. We can add more, remove or modify the files in this folder
+- Chart.yaml: this file is the description about the chart information
+- _helpers.tpl: this file defines some yaml key-value which we can import to yaml files in templates folders
+
+Let take a look about the file template/deployments.yaml:
+
+```yaml
+
+  apiVersion: apps/v1
+  kind: Deployment
+  metadata:
+    name: {{ include "mynginx.fullname" . }}
+    labels:
+      {{- include "mynginx.labels" . | nindent 4 }}
+  spec:
+    {{- if not .Values.autoscaling.enabled }}
+    replicas: {{ .Values.replicaCount }}
+    {{- end }}
+    selector:
+      matchLabels:
+        {{- include "mynginx.selectorLabels" . | nindent 6 }}
+    template:
+      metadata:
+        {{- with .Values.podAnnotations }}
+        annotations:
+          {{- toYaml . | nindent 8 }}
+        {{- end }}
+        labels:
+          {{- include "mynginx.selectorLabels" . | nindent 8 }}
+      spec:
+        {{- with .Values.imagePullSecrets }}
+        imagePullSecrets:
+          {{- toYaml . | nindent 8 }}
+        {{- end }}
+        serviceAccountName: {{ include "mynginx.serviceAccountName" . }}
+        securityContext:
+          {{- toYaml .Values.podSecurityContext | nindent 8 }}
+        containers:
+          - name: {{ .Chart.Name }}
+            securityContext:
+              {{- toYaml .Values.securityContext | nindent 12 }}
+            image: "{{ .Values.image.repository }}:{{ .Values.image.tag | default .Chart.AppVersion }}"
+            imagePullPolicy: {{ .Values.image.pullPolicy }}
+  ..........................................................
+
+```
+
+This yaml file looks like the template of k8s resource. But it has some {{ }}, the statement in {{ }} can be a value, operator, expression,.... Let me explain some popular thing:
+
+- {{ include … }}: this syntax help us to import the yaml template that is defined in file **_helpers.tpl**
+- {{ .Values.&#60;key&#62; }}: get the **value** at the position **key** of files values.yaml and fill that value to the k8s template yaml file
+- {{ -if &#60;expression &#62; }} &#60;some yaml line&#62; {{- end }}:  check if the expression return true, the &#60;some yaml line&#62; will be added to the template files
+- {{ with .Values.key }} {{- toYaml . }}   {{- end }}, {{- toYaml .Values.key }}: get the **value** at the position **key** of files **values.yaml** and convert it to yaml format.
```

Comment:
}}:  if the expression returns true, , the &#60;some yaml line&#62; will be added to the template file